lib/rules-generator.js

```diff
@@ -187,93 +203,102 @@ util.methods(Generator, {
                                 [ ast.call(ast.variable('@RegExp'), [ast.variable('s')]) ])),
     });
 
-    registerAsCall('Object', 'hasChildren');
     registerAsCall('Number', 'isNumber');
     registerAsCall('Boolean', 'isBoolean');
+
+    this.symbols.registerFunction('@RegExp', ['s'],
+                                  ast.builtin(this.makeRegExp.bind(this)));
+  },
+
+  // Ensure we have a definition for a validator for the given schema.
+  ensureValidator: function(schemaName) {
+    // TODO: Guard against recursion
+    if (!this.validators[schemaName]) {
+      this.validators[schemaName] = this.createValidator(schemaName);
+      // console.log(schemaName + ": " + util.prettyJSON(this.validators[schemaName]));
+    }
+    return this.validators[schemaName];
   },
 
-  // Return a validator expression that returns true iff
-  // *this* conforms to the defined schema:
-  // 1. Conforms to derivedFrom schema. &&
-  // 2. Each property in properties conforms to it's type.
-  // 3. Conforms to the validate method.
-  registerSchemaValidator: function(schemaName) {
+  // A validator is a structured object, where each leaf node
+  // is:
+  //     ".validate": [<expression>, ...]
+  // All expressions will be ANDed together to form the file expresion.
+  // Intermediate nodes can be "prop" or "$prop" values.
+  createValidator: function(schemaName) {
     var schema = this.symbols.schema[schemaName];
-    var terms = [];
-    var thisVar = ast.variable('this');
+    var validator = {};
+
+    if (!schema) {
+      throw new Error(errors.application + "Undefined schema: " + schemaName);
+    }
 
     var hasProps = Object.keys(schema.properties).length > 0;
 
     if (hasProps && !this.symbols.isDerivedFrom(schemaName, 'Object')) {
-      this.fatal(errors.nonObject + " (" + schemaName + " is " + schema.derivedFrom + ")");
-      return;
+      this.fatal(errors.nonObject + " (" + schemaName + " is " + schema.derivedFrom.name + ")");
+      return {};
     }
 
-    // @validator@<T>(this)
-    if (schema.derivedFrom != 'Any' && !hasProps) {
-      terms.push(ast.call(ast.variable('@validator@' + schema.derivedFrom),
-                          [ thisVar ]));
+    if (schema.derivedFrom.name != 'Any') {
+      extendValidator(validator, this.ensureValidator(schema.derivedFrom.name));
     }
 
-    for (var propName in schema.properties) {
-      var property = schema.properties[propName];
-      var propTerms = [];
-      for (var i = 0; i < property.types.length; i++) {
-        var type = property.types[i];
-        // @validator@<type>(this[propName])
-        propTerms.push(ast.call(ast.variable('@validator@' + type),
-                                [ ast.reference(thisVar, propName) ]));
+    var requiredProperties = [];
+    Object.keys(schema.properties).forEach(function(propName) {
```

Comment:
IFAIK, forEach is only available for Array, not Object.  So using Object.keys() to get the keys as an array, and then enumerator over those.  It IS kind of gross.  I could add a objectForEach() type enumerator either from an npm package (overkill) or just add to util.js locally here.